### PR TITLE
Block admission until PodsReady for all admitted workloads if WaitForPodsReady is enabled

### PR DESCRIPTION
**Pull request by mimowo**
#### What type of PR is this?

/kind feature

#### What this PR does / why we need it:

Introduces sequencing of job startups to prevent deadlocks in case of resource provisioning issues.

#### Which issue(s) this PR fixes:

Part of: https://github.com/kubernetes-sigs/kueue/issues/349

#### Special notes for your reviewer:

- built on top of the current version of https://github.com/kubernetes-sigs/kueue/pull/460, will rebase once the first PR is merged

### CODE REVIEW

File: pkg/scheduler/scheduler.go [hunk 1]
```diff
@@ -124,6 +154,9 @@ func (s *Scheduler) schedule(ctx context.Context) {
 		log := log.WithValues("workload", klog.KObj(e.Obj), "clusterQueue", klog.KRef("", e.ClusterQueue))
 		if err := s.admit(ctrl.LoggerInto(ctx, log), e); err != nil {
 			e.inadmissibleMsg = fmt.Sprintf("Failed to admit workload: %v", err)
+		} else if s.waitForPodsReady {
+			klog.InfoS("Workload successfully admitted when waitForPodsReady - do admit more in this cycle", "workload", e.Obj.Name)
```
Comment (alculquicondor) [inline pkg/scheduler/scheduler.go:158]: halting more admissions in this cycle.

Admit should already log the name of the workload, so I don't think we should include it again.

Comment (mimowo) [reply pkg/scheduler/scheduler.go]: That's right, cleaned up

File: pkg/cache/cache.go [hunk 2]
```diff
@@ -562,6 +651,7 @@ func (c *Cache) UpdateWorkload(oldWl, newWl *kueue.Workload) error {
 	if !ok {
 		return fmt.Errorf("new ClusterQueue doesn't exist")
 	}
+	c.broadcastToRecheckPodsReadyForUpdatedWorkload(oldWl, newWl)
```
[the remark this answers is not in the document]
Comment (mimowo) [reply pkg/cache/cache.go:655]: I want to broadcast only if it makes sense to recheck the condition - meaning that there was an admitted workload with PodsReady=False and now the workload has PodsReady=True or is gone. Rechecking the condition on adding a new workload does not seem necessary.

File: pkg/cache/cache.go [hunk 3]
```diff
@@ -148,6 +177,66 @@ func (c *Cache) newClusterQueue(cq *kueue.ClusterQueue) (*ClusterQueue, error) {
...
+			c.cond.Wait()
+		}
+	}
+}
+
+func (c *Cache) podsReadyForAllAdmittedWorkloads() bool {
+	workloadsToCheck := []*kueue.Workload{}
+	for _, cq := range c.clusterQueues {
+		for _, wl := range cq.Workloads {
+			workloadsToCheck = append(workloadsToCheck, wl.Obj)
```
[the remark this answers is not in the document]
Comment (mimowo) [reply pkg/cache/cache.go:204]: It was written this way for debugging purposes, I will follow up with a clean up break early if there is a workload with PodsReady=False.

File: pkg/scheduler/scheduler.go [hunk 4]
```diff
@@ -124,6 +154,9 @@ func (s *Scheduler) schedule(ctx context.Context) {
 		log := log.WithValues("workload", klog.KObj(e.Obj), "clusterQueue", klog.KRef("", e.ClusterQueue))
 		if err := s.admit(ctrl.LoggerInto(ctx, log), e); err != nil {
```
[the remark this answers is not in the document]
Comment (mimowo) [reply pkg/scheduler/scheduler.go:155]: The reasoning to block early in this function was that the blocking may take a while (say a couple of minutes) and after the time the nominated workloads could be different. However, this seems a minor concern. 

I think that another advantage of the suggested approach is that this way it will be easier to implement the more granular version of the feature with blocking based on overlapping resource flavors. Going to follow up with a commit along these lines.

[the remark this answers is not in the document]
Comment (alculquicondor) [reply pkg/scheduler/scheduler.go:155]: You make a good point. Workloads could have finished and freed resources. But at least we know that this workload would fit.

File: main.go [hunk 5]
```diff
@@ -111,7 +111,7 @@ func main() {
 		close(certsReady)
 	}
 
-	cCache := cache.New(mgr.GetClient())
+	cCache := cache.New(mgr.GetClient(), cache.WithWaitForPodsReady(waitForPodsReady(&cfg)))
```
[the remark this answers is not in the document]
Comment (mimowo) [reply main.go]: renamed the option as `podsReadyTracking`.

File: pkg/cache/cache.go [hunk 6]
```diff
@@ -148,6 +177,66 @@ func (c *Cache) newClusterQueue(cq *kueue.ClusterQueue) (*ClusterQueue, error) {
...
+			return
+		}
+		log.V(3).Info("Blocking admission as not all workloads are in the PodsReady condition")
+		select {
+		case <-ctx.Done():
+			return
+		default:
+			c.cond.Wait()
+		}
+	}
```
[the remark this answers is not in the document]
Comment (mimowo) [reply pkg/cache/cache.go]: Applied the same pattern as for the manager.

File: pkg/cache/cache.go [hunk 7]
```diff
@@ -148,6 +177,66 @@ func (c *Cache) newClusterQueue(cq *kueue.ClusterQueue) (*ClusterQueue, error) {
...
+
+func (c *Cache) broadcastToRecheckPodsReady() {
+	c.cond.Broadcast()
+}
+
+func (c *Cache) broadcastToRecheckPodsReadyForDeletedWorkload(w *kueue.Workload) {
+	if c.waitForPodsReady {
+		// deleting a PodsReady=false workload
+		if !workload.InCondition(w, kueue.WorkloadPodsReady) {
+			c.broadcastToRecheckPodsReady()
```
[the remark this answers is not in the document]
Comment (mimowo) [reply pkg/cache/cache.go]: Inlined

File: pkg/cache/cache.go [hunk 8]
```diff
@@ -578,6 +668,7 @@ func (c *Cache) DeleteWorkload(w *kueue.Workload) error {
 	}
 
 	c.cleanupAssumedState(w)
+	c.broadcastToRecheckPodsReadyForDeletedWorkload(w)
```
Comment (ahg-g) [inline pkg/cache/cache.go:671]: the function is short and its name is not too friendly, I would inline it:

```
c.podsReadyTracking && !apimeta.IsStatusConditionTrue(w.Status.Conditions, kueue.WorkloadPodsReady) {
  c.podsReadyCond.Broadcast()
}
```

File: pkg/cache/cache.go [hunk 9]
```diff
@@ -627,6 +718,7 @@ func (c *Cache) ForgetWorkload(w *kueue.Workload) error {
 		return errCqNotFound
 	}
 	cq.deleteWorkload(w)
+	c.broadcastToRecheckPodsReadyForDeletedWorkload(w)
```
Comment (ahg-g) [inline pkg/cache/cache.go:721]: ditto

File: pkg/cache/cache.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (ahg-g) [reply pkg/cache/cache.go:654]: +1 to inlining here

File: pkg/cache/cache.go [hunk 10]
```diff
@@ -148,6 +178,65 @@ func (c *Cache) newClusterQueue(cq *kueue.ClusterQueue) (*ClusterQueue, error) {
...
+		default:
+			c.podsReadyCond.Wait()
+		}
+	}
+}
+
+func (c *Cache) podsReadyForAllAdmittedWorkloads(ctx context.Context) bool {
+	log := ctrl.LoggerFrom(ctx)
+	for _, cq := range c.clusterQueues {
+		for _, wl := range cq.Workloads {
```
Comment (ahg-g) [inline pkg/cache/cache.go:203]: Something to think about: what if we track "not ready" workloads in each ClusterQueue, and so instead of iterating over all workloads every time, we just check if that list is empty? This will also make it not expensive to call `podsReadyForAllAdmittedWorkloads`, and so we can remove the checks that we have around broadcast.

Comment (alculquicondor) [reply pkg/cache/cache.go:203]: +1, although I'm fine leaving this as a follow up, to focus on the semantics first.

File: pkg/cache/cache.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (alculquicondor) [reply pkg/cache/cache.go:654]: Back to this:

> Also you probably need to check when the Workload is first added

when we admit a workload, it could immediately start running and become ready before kueue gets the event that runs `AddOrUpdateWorkload`. So you need to check if the pods are ready and broadcast in that function too.

File: pkg/cache/cache_test.go [hunk 11]
```diff
@@ -1733,6 +1733,44 @@ func TestClusterQueueUpdateCodependentResources(t *testing.T) {
 	}
 }
 
+// TestHeadsCancelled ensures that the Heads call returns when the context is closed.
+func TestHeadsCancelled(t *testing.T) {
```
Comment (alculquicondor) [inline pkg/cache/cache_test.go:1737]:
```suggestion
func TestWaitForPodsReadyCancelled(t *testing.T) {
```

Although the test looks just copied from the next one?

File: pkg/scheduler/scheduler.go [hunk 12]
```diff
@@ -121,9 +144,19 @@ func (s *Scheduler) schedule(ctx context.Context) {
...
+			// Block admission until all currently admitted workloads are in
+			// PodsReady condition if the waitForPodsReady is enabled
+			s.cache.WaitForPodsReady(ctx)
+		}
+
 		log := log.WithValues("workload", klog.KObj(e.Obj), "clusterQueue", klog.KRef("", e.ClusterQueue))
 		if err := s.admit(ctrl.LoggerInto(ctx, log), e); err != nil {
 			e.inadmissibleMsg = fmt.Sprintf("Failed to admit workload: %v", err)
+		} else if s.waitForPodsReady {
+			klog.InfoS("Workload successfully admitted when waitForPodsReady - do admit more in this cycle")
```
Comment (alculquicondor) [inline pkg/scheduler/scheduler.go:158]: do not?

Comment (mimowo) [reply pkg/scheduler/scheduler.go:158]: Yes, but I think this if is no longer necessary once we moved the waiting inside the loop over heads.

File: pkg/cache/cache_test.go (hunk 11 above)
Comment (mimowo) [reply pkg/cache/cache_test.go]: That's right, now fixed.

File: pkg/cache/cache.go (hunk 9 above)
Comment (mimowo) [reply pkg/cache/cache.go]: cleaned up

File: pkg/cache/cache.go (hunk 8 above)
Comment (mimowo) [reply pkg/cache/cache.go]: I have inlined and just check `podsReadyTracking` since checking the condition in `podsReadyForAllAdmittedWorkloads`  should now be fast after caching the WorkloadsWithPodsReady at the ClusterQueue level.

File: pkg/cache/cache.go [hunk 13]
```diff
@@ -148,6 +178,65 @@ func (c *Cache) newClusterQueue(cq *kueue.ClusterQueue) (*ClusterQueue, error) {
...
+	for {
+		if c.podsReadyForAllAdmittedWorkloads(ctx) {
+			return
+		}
+		log.V(3).Info("Blocking admission as not all workloads are in the PodsReady condition")
+		select {
+		case <-ctx.Done():
+			return
+		default:
+			c.podsReadyCond.Wait()
```
[the remark this answers is not in the document]
Comment (mimowo) [reply pkg/cache/cache.go:197]: Done

File: pkg/cache/cache.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (mimowo) [reply pkg/cache/cache.go]: Done

File: pkg/scheduler/scheduler.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (mimowo) [reply pkg/scheduler/scheduler.go:155]: Moved the waiting just before admit

File: pkg/cache/cache.go (hunk 10 above)
Comment (mimowo) [reply pkg/cache/cache.go]: Done, I think it also simplifies the code.

File: pkg/scheduler/scheduler.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (mimowo) [reply pkg/scheduler/scheduler.go:163]: > Also, it might be useful to produce a Workload status update if the Workload satisfies the quota, but is blocked waiting on others Workloads to be ready.

Now, I set the `WorkloadAdmitted=False` condition with `Reason=Waiting`. It allows me to wait for the situation in the integration test.

File: pkg/cache/cache.go [hunk 14]
```diff
@@ -148,6 +179,53 @@ func (c *Cache) newClusterQueue(cq *kueue.ClusterQueue) (*ClusterQueue, error) {
...
+	c.Lock()
+	defer c.Unlock()
+	return c.podsReadyForAllAdmittedWorkloads(ctx)
+}
+
+func (c *Cache) podsReadyForAllAdmittedWorkloads(ctx context.Context) bool {
+	log := ctrl.LoggerFrom(ctx)
+	for _, cq := range c.clusterQueues {
+		if len(cq.WorkloadsWithoutPodsReady) > 0 {
+			log.V(3).Info("There is a cluster queue with not ready workloads", "clusterQueue", cq.Name, "workloads", cq.WorkloadsWithoutPodsReady.List())
```
[the remark this answers is not in the document]
Comment (alculquicondor) [reply pkg/cache/cache.go:213]: also, wrap it with `if logV := log.V(3); logV.Enabled() { }`, otherwise we pay the cost of calling `List` and serializing every time.
Or just remove the list... I don't think it will be that useful.

File: pkg/cache/cache.go [hunk 15]
```diff
@@ -148,6 +179,53 @@ func (c *Cache) newClusterQueue(cq *kueue.ClusterQueue) (*ClusterQueue, error) {
...
+
+	log := ctrl.LoggerFrom(ctx)
+	for {
+		if c.podsReadyForAllAdmittedWorkloads(ctx) {
+			return
+		}
+		log.V(3).Info("Blocking admission as not all workloads are in the PodsReady condition")
+		select {
+		case <-ctx.Done():
+			log.V(5).Info("Context cancelled, returning for the waiting")
```
Comment (alculquicondor) [inline pkg/cache/cache.go:194]:
```suggestion
			log.V(5).Info("Context cancelled when waiting for pods to be ready; returning")
```

File: pkg/cache/cache.go [hunk 16]
```diff
@@ -263,6 +341,9 @@ func (c *ClusterQueue) addWorkload(w *kueue.Workload) error {
 	wi := workload.NewInfo(w)
 	c.Workloads[k] = wi
 	c.updateWorkloadUsage(wi, 1)
+	if !apimeta.IsStatusConditionTrue(w.Status.Conditions, kueue.WorkloadPodsReady) {
```
Comment (alculquicondor) [inline pkg/cache/cache.go:344]: also, if `c.podsReadyTracking`

File: pkg/cache/cache.go [hunk 17]
```diff
@@ -627,6 +720,7 @@ func (c *Cache) ForgetWorkload(w *kueue.Workload) error {
 		return errCqNotFound
 	}
 	cq.deleteWorkload(w)
+	c.podsReadyCond.Broadcast()
```
Comment (alculquicondor) [inline pkg/cache/cache.go:723]: `if c.podsReadyTracking`

File: pkg/cache/cache_test.go [hunk 18]
```diff
@@ -1770,6 +1771,303 @@ func TestMatchingClusterQueues(t *testing.T) {
...
+
+	if cache.podsReadyForAllAdmittedWorkloads(ctx) {
+		t.Fatalf("Unexpected that all admitted workloads are in PodsReady condition")
+	}
+
+	// cancel the context so that the WaitForPodsReady is returns
+	go func() {
+		cancel()
+		cache.CleanUpOnContext(ctx)
+	}()
```
Comment (alculquicondor) [inline pkg/cache/cache_test.go:1807]:
```suggestion
	go cancel()
```

`CleanUpOnContext` could be called before.

File: pkg/cache/cache.go [hunk 19]
```diff
@@ -274,6 +355,9 @@ func (c *ClusterQueue) deleteWorkload(w *kueue.Workload) {
 		return
 	}
 	c.updateWorkloadUsage(wi, -1)
+	if !apimeta.IsStatusConditionTrue(w.Status.Conditions, kueue.WorkloadPodsReady) {
+		c.WorkloadsWithoutPodsReady.Delete(k)
```
Comment (alculquicondor) [inline pkg/cache/cache.go:359]: what about calling broadcast here?

File: test/integration/scheduler/podsready/scheduler_test.go [hunk 20]
```diff
@@ -0,0 +1,133 @@
...
+	)
+
+	ginkgo.BeforeEach(func() {
+		ns = &corev1.Namespace{
+			ObjectMeta: metav1.ObjectMeta{
+				GenerateName: "core-",
+			},
+		}
+		gomega.Expect(k8sClient.Create(ctx, ns)).To(gomega.Succeed())
+		defaultFlavor = testing.MakeResourceFlavor("default").Label(instanceKey, "default").Obj()
```
Comment (alculquicondor) [inline test/integration/scheduler/podsready/scheduler_test.go:51]: To simplify the test:

```suggestion
		defaultFlavor = testing.MakeResourceFlavor("default").Obj()
```

File: test/integration/scheduler/podsready/scheduler_test.go [hunk 21]
```diff
@@ -0,0 +1,133 @@
...
+		ginkgo.It("Should unblock admission of new workloads once the admitted workload is in PodsReady condition", func() {
+			ginkgo.By("checking the first prod workload gets admitted")
+			prodWl := testing.MakeWorkload("prod-wl", ns.Name).Queue(prodQueue.Name).Request(corev1.ResourceCPU, "2").Obj()
+			gomega.Expect(k8sClient.Create(ctx, prodWl)).Should(gomega.Succeed())
+			framework.ExpectWorkloadsToBeAdmitted(ctx, k8sClient, prodClusterQ.Name, prodWl)
+
+			ginkgo.By("verifying that the second workload is waiting as the first one has PodsReady=False")
+			devWl := testing.MakeWorkload("dev-wl", ns.Name).Queue(devQueue.Name).Request(corev1.ResourceCPU, "2").Obj()
+			gomega.Expect(k8sClient.Create(ctx, devWl)).Should(gomega.Succeed())
+			framework.ExpectWorkloadsToBeWaiting(ctx, k8sClient, devWl)
```
Comment (alculquicondor) [inline test/integration/scheduler/podsready/scheduler_test.go:102]: I think this test is too gentle:

```suggestion
			devWl := testing.MakeWorkload("dev-wl", ns.Name).Queue(devQueue.Name).Request(corev1.ResourceCPU, "2").Obj()
			gomega.Expect(k8sClient.Create(ctx, devWl)).Should(gomega.Succeed())
			framework.ExpectWorkloadsToBeAdmitted(ctx, k8sClient, prodClusterQ.Name, prodWl)
			framework.ExpectWorkloadsToBeWaiting(ctx, k8sClient, devWl)
```

Basically, create the workloads in rapid succession to increase the chances of race conditions.

In particular, this would ensure that `AssumeWorkload` also puts the workload in the list of unready workloads (which I think it does already).

File: test/integration/scheduler/podsready/scheduler_test.go (hunk 20 above)
Comment (alculquicondor) [reply test/integration/scheduler/podsready/scheduler_test.go:51]: Also, you can create the RF once for the suite in `BeforeSuite`. We don't really need to be recreating it, making our tests cases a bit slower.

File: test/integration/scheduler/podsready/scheduler_test.go (hunk 21 above)
Comment (mimowo) [reply test/integration/scheduler/podsready/scheduler_test.go:104]: ack, also just added a new test which makes sure both workloads are considered in the same scheduling cycle

File: pkg/cache/cache.go (hunk 19 above)
Comment (mimowo) [reply pkg/cache/cache.go:359]: I was thinking about it, it would work. The only reason I haven't done this is performance consideration. For example, when UpdateWorkload is called it would result in two broadcasts - one after delete and one after add. But maybe this is not a concern, especially as now checking the condition is faster, what do you think?

File: pkg/cache/cache.go (hunk 17 above)
Comment (mimowo) [reply pkg/cache/cache.go]: fixed

File: pkg/cache/cache.go (hunk 16 above)
Comment (mimowo) [reply pkg/cache/cache.go]: we are here in cluster queue, so also needed to pass the `podsReadyTracking` down from cache.

File: pkg/cache/cache_test.go (hunk 18 above)
Comment (mimowo) [reply pkg/cache/cache_test.go]: done

File: pkg/cache/cache.go (hunk 14 above)
[the remark this answers is not in the document]
Comment (mimowo) [reply pkg/cache/cache.go]: yup, no longer logging it

File: pkg/cache/cache.go [hunk 22]
```diff
@@ -148,6 +179,53 @@ func (c *Cache) newClusterQueue(cq *kueue.ClusterQueue) (*ClusterQueue, error) {
 	return cqImpl, nil
 }
 
+func (c *Cache) WaitForPodsReady(ctx context.Context) {
```
[the remark this answers is not in the document]
Comment (mimowo) [reply pkg/cache/cache.go:186]: just returning. Returning makes it possible to test what happens

File: test/integration/scheduler/podsready/scheduler_test.go (hunk 20 above)
Comment (mimowo) [reply test/integration/scheduler/podsready/scheduler_test.go:51]: Hm, I'm wondering about the new context instead of suite though, I have two concerns:
- the framework is currently setup in BeforeSuite, so it would require refactoring to move the configuration from `suite_test.go` to `scheduler_test.go`, then it no longer can be BeforeSuite but rather BeforeAll or BeforeEach.
- new context means that tests within the suite can no longer execute in parallel (do we support [Serial] tag in Kueue?), it may actually result in worse performance.

File: test/integration/scheduler/podsready/scheduler_test.go (hunk 21 above)
Comment (mimowo) [reply test/integration/scheduler/podsready/scheduler_test.go]: done

File: pkg/cache/cache_test.go [hunk 23]
```diff
@@ -1770,6 +1771,302 @@ func TestMatchingClusterQueues(t *testing.T) {
...
+			}
+			if tc.setup != nil {
+				if err := tc.setup(cache); err != nil {
+					t.Errorf("Unexpected error during setup: %q", err)
+				}
+			}
+			if err := tc.operation(cache); err != nil {
+				t.Errorf("Unexpected error during operation: %q", err)
+			}
+			gotPodsReadyForAllAdmittedWorkloads := cache.podsReadyForAllAdmittedWorkloads(ctx)
```
Comment (alculquicondor) [inline pkg/cache/cache_test.go:2058]:
```suggestion
			gotPodsReadyForAllAdmittedWorkloads := cache.PodsReadyForAllAdmittedWorkloads(ctx)
```

File: test/integration/scheduler/podsready/scheduler_test.go [hunk 24]
```diff
@@ -0,0 +1,156 @@
...
+
+		ginkgo.It("Should block admission of one new workload if two are considered in the same scheduling cycle", func() {
+			ginkgo.By("creating two workloads but delaying cluster queue creation which has enough capacity")
+			prodWl := testing.MakeWorkload("prod-wl", ns.Name).Queue(prodQueue.Name).Request(corev1.ResourceCPU, "11").Obj()
+			gomega.Expect(k8sClient.Create(ctx, prodWl)).Should(gomega.Succeed())
+			devWl := testing.MakeWorkload("dev-wl", ns.Name).Queue(devQueue.Name).Request(corev1.ResourceCPU, "11").Obj()
+			gomega.Expect(k8sClient.Create(ctx, devWl)).Should(gomega.Succeed())
+			framework.ExpectWorkloadsToBePending(ctx, k8sClient, prodWl, devWl)
+
+			ginkgo.By("creating the cluster queue to ensure the two workloads are considered in the same scheduling cycle")
```
Comment (alculquicondor) [inline test/integration/scheduler/podsready/scheduler_test.go:135]: The rest is already explained in the first `By` and in the comment

```suggestion
			ginkgo.By("creating the cluster queue")
```

File: test/integration/scheduler/podsready/scheduler_test.go (hunk 20 above)
Comment (alculquicondor) [reply test/integration/scheduler/podsready/scheduler_test.go]: Ah right... we share the same controllers across the suite.

Separate suite makes sense. I didn't know about `BeforeAll`, we should be using it in other tests. And we should create the common RF either `BeforeSuite` or `BeforeAll`. But that's a nit.

**Review by alculquicondor — COMMENTED** [review]
/approve
Just nits left

You can squash as you address them.

File: pkg/scheduler/scheduler.go [hunk 25]
```diff
@@ -121,6 +144,21 @@ func (s *Scheduler) schedule(ctx context.Context) {
...
 		}
+
+		if s.waitForPodsReady {
+			if !s.cache.PodsReadyForAllAdmittedWorkloads(ctx) {
+				log.V(5).Info("Waiting for all admitted workloads to be in the PodsReady condition")
+				// Block admission until all currently admitted workloads are in
+				// PodsReady condition if the waitForPodsReady is enabled
+				if err := workload.UpdateStatus(ctx, s.client, e.Obj, kueue.WorkloadAdmitted, metav1.ConditionFalse, "Waiting", "waiting for all admitted workloads to be in PodsReady condition"); err != nil {
+					log.Error(err, "Could not update Workload status")
+					continue
```
Comment (ahg-g) [inline pkg/scheduler/scheduler.go:155]: why `continue` here? failing to update the status here isn't critical, we shouldn't give up on admitting the workload here.

File: pkg/cache/cache.go [hunk 26]
```diff
@@ -94,12 +123,13 @@ const (
...
-	RequestableResources map[corev1.ResourceName]*Resource
-	UsedResources        ResourceQuantities
-	Workloads            map[string]*workload.Info
-	NamespaceSelector    labels.Selector
+	Name                      string
+	Cohort                    *Cohort
+	RequestableResources      map[corev1.ResourceName]*Resource
+	UsedResources             ResourceQuantities
+	Workloads                 map[string]*workload.Info
+	WorkloadsWithoutPodsReady sets.String
```
Comment (ahg-g) [inline pkg/cache/cache.go:131]: I am trying to find a better name, this is hard to parse, may be `NotReadyWorkloads` or `WorkloadsNotReady` to keep Workloads are the prefix?

Comment (mimowo) [reply pkg/cache/cache.go:131]: I like `WorkloadsNotReady`, renamed

File: pkg/scheduler/scheduler.go (hunk 25 above)
Comment (mimowo) [reply pkg/scheduler/scheduler.go:163]: My reasoning was that if the update fails it is likely due to conflicts as `e` holds an outdated workload state, so the actual admission is also likely to fail.

However, we need to be prepared anyway to handle outdated workload state during admission anyway, so we should be ok with trying admission (and it may actually succeed.  Reverting the `continue`.

File: pkg/cache/cache_test.go [hunk 27]
```diff
@@ -1770,6 +1771,302 @@ func TestMatchingClusterQueues(t *testing.T) {
...
+			if diff := cmp.Diff(tc.wantPodsReadyForAllAdmittedWorkloads, gotPodsReadyForAllAdmittedWorkloads); diff != "" {
```
[the remark this answers is not in the document]
Comment (mimowo) [reply pkg/cache/cache_test.go:2059]: Right, renamed

File: pkg/cache/cache_test.go (hunk 23 above)
Comment (mimowo) [reply pkg/cache/cache_test.go:2058]: Done

File: test/integration/scheduler/podsready/scheduler_test.go (hunk 20 above)
Comment (mimowo) [reply test/integration/scheduler/podsready/scheduler_test.go]: Moved the RF setup to `BeforeSuite`, but it does not make any noticeable speedup on my local machine.
I think `BeforeAll` might be tricky as it works, but requires the container to be ordered, then all tests within the container are executed sequentially, which in our case might not be necessary.

File: pkg/cache/cache.go (hunk 19 above)
Comment (mimowo) [reply pkg/cache/cache.go:367]: we could pass the broadcast function down, but not sure it is the right approach indeed. Keeping as is for now.

**Comment by mimowo** [comment]
Applied the remarks and squashed the commits. PTAL

**Comment by ahg-g** [comment]
/lgtm

Great!
